ui/src/components/LocationFieldset.js

```diff
@@ -0,0 +1,231 @@
+        + 'known location of a person.'),
+  },
+  showMap: {
+    id: 'LocationFieldset.showMap',
+    defaultMessage: 'Show map',
+    description: 'Label on a button to show a map.',
+  },
+});
+
+const BASE_MAPS_API_URL = 'https://maps.googleapis.com/maps/api/js';
+
+class LocationFieldset extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      showMap: false,
+      haveStartedLoadingMapScript: false,
+      haveFinishedLoadingMapScript: false,
+    };
+    this.mapsApiDisabled = ENV.maps_api_key == null || ENV.maps_api_key == '';
+    this.onLocationLatLngUpdate = this.onLocationLatLngUpdate.bind(this);
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    // This is more complex than I'd like, but basically what's going on is I
+    // want to ensure the Google Maps JS is never loaded more than once, and
+    // only loaded at all if the user wants to show the map.
+    // The first time showMap is set to true, we also set
+    // haveStartedLoadingMapScript to true, and when that's set to true we load
+    // the map script.
+    if (this.state.showMap && !prevState.showMap) {
+      this.setState({haveStartedLoadingMapScript: true});
+    } else if (this.state.haveStartedLoadingMapScript &&
+        !prevState.haveStartedLoadingMapScript) {
+      Utils.loadExternalScript(
+        BASE_MAPS_API_URL + '?key=' + ENV.maps_api_key,
+        () => this.setState({haveFinishedLoadingMapScript: true}));
+    }
+  }
+
+  onLocationLatLngUpdate(value) {
+    const textValue = value[0] + ', ' + value[1];
+    this.props.onLocationLatLngUpdate(value);
+    this.props.onLocationTextUpdate(textValue);
+  }
+
+  onLocationTextFieldBlur(e) {
+    if (this.state.haveFinishedLoadingMapScript) {
+      const geocoder = new google.maps.Geocoder();
+      geocoder.geocode(
+          {address: this.props.locationText},
+          (results, status) => {
+            if ((status != 'OK') || (results.length == 0)) {
+              // We're geocoding strictly on a best-effort basis.
+              return;
+            }
+            const resultGeoLocation = results[0].geometry.location;
+            this.props.onLocationLatLngUpdate(
+                [resultGeoLocation.lat(), resultGeoLocation.lng()]);
+          });
+    }
+  }
+
+  render() {
+    var showHideMapButton = null;
+    if (!this.mapsApiDisabled) {
+      showHideMapButton = this.state.showMap ?
+          (
+            <Button
+              className='pf-button-primary'
+              type='button'
+              onClick={() => this.setState({showMap: false})}>
+              {this.props.intl.formatMessage(MESSAGES.hideMap)}
+            </Button>
+          ) :
+          (
+            <Button
+              className='pf-button-primary'
+              type='button'
+              onClick={() => this.setState({showMap: true})}>
+              {this.props.intl.formatMessage(MESSAGES.showMap)}
+            </Button>
```

Comment:
Can you reduce the duplication here? Something like:

```js
const newMapState = !this.state.showMap;
showHideMapButton = (<Button>...onclick={() => this.setState({showMap: newMapState})...</Button>);
```

Reply:
Done.